### PR TITLE
fix: compactor loop in highly backlogged case

### PR DESCRIPTION
**Pull request by Joe-Blount**
Closes https://github.com/influxdata/idpe/issues/17896

I noticed a partition stuck looping on compactions, it had just under 1900 L0 files.  I suspect if it had kept caught up all along it never would have gotten stuck looping with unproductive compactions.  The scenario seems to be produced by it catching up from a highly backlogged state.  The test case added in this PR replicates what I believe is the underlying problem.  The change made in this PR fixes the added test case, and the ugly test case generated from the 1900 L0 files observed.

I am not checking in the test case from the original scenario observed in the catalog.  Creating a test with 1900 specific files is over 23,000 lines of code.  In addition to it being a huge amount of code no one can realistically read, rebuilding the file after a tweak (including after insta updates) stalls my system for minutes.  That was a necessary step to get the knowledge to create a focused test.  And I'm happy to delete it :) 

All that leads to a small simple change, that be described as:  When a large set of overlapping L0s meets the criteria for `ManySmallFiles` (which does L0->L0 compactions), we should not assume that all files within the set are small and need L0->L0 compactions.  So a check is added to identify groups of files within that set that happen to large.  If such large files are found, they're set aside for later, when we're not in the `ManySmallFiles` mode, at which point we'll compact them to L1, which is what they're ready for.

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: compactor/tests/layouts/many_files.rs [hunk 1]
```diff
@@ -4670,17 +4670,17 @@ async fn l0s_almost_needing_vertical_split() {
     - "L0.998[24,100] 1.02us    |-----------------------------------------L0.998-----------------------------------------|"
     - "L0.999[24,100] 1.02us    |-----------------------------------------L0.999-----------------------------------------|"
     - "L0.1000[24,100] 1.02us   |----------------------------------------L0.1000-----------------------------------------|"
-    - "**** Final Output Files (7.41gb written)"
+    - "**** Final Output Files (6.75gb written)"
```
Comment (NGA-TRAN) [inline compactor/tests/layouts/many_files.rs:4673]: I wonder why the size is different

File: compactor/src/components/file_classifier/split_based.rs [hunk 2]
```diff
@@ -199,12 +202,21 @@ fn file_classification_for_many_files(
...
             .chunks(max_num_files_to_group)
             .for_each(|chunk| {
-                if files_to_compact.is_empty() {
+                let this_chunk_bytes: usize =
+                    chunk.iter().map(|f| f.file_size_bytes as usize).sum();
+                if this_chunk_bytes > max_total_file_size_to_group {
+                    // This chunk of files are plenty big and don't fit the ManySmallFiles charactoristics.
+                    // If we let ManySmallFiles handle them, it may get stuck with unproductive compactions.
+                    // So set them aside for later (when we're not in ManySmallFiles mode).
+                    files_to_keep.append(chunk.to_vec().as_mut());
```
Comment (NGA-TRAN) [inline compactor/src/components/file_classifier/split_based.rs:215]: How many rounds these files then will get compacted into L1s? If it is one round, then I am afraid we will hit the same issue that too-many files and OOM.
I wonder whether all algorithms we modified so far have helped making these number of files small?

**Review by NGA-TRAN — APPROVED** [review]
I am a bit suspicious and have questions but I trust your instinct and we can always revert this PR if it does not work.

On a different but related note. I wonder if we can have a warning on the dashboard to tell us there are too many L0 files in a partition. Just like a panic warning. I am not sure how tricky to do this though

File: compactor/src/components/file_classifier/split_based.rs [hunk 3]
```diff
@@ -199,12 +202,21 @@ fn file_classification_for_many_files(
...
         let ordered_files = order_files(files, target_level.prev());
+
         ordered_files
             .chunks(max_num_files_to_group)
             .for_each(|chunk| {
-                if files_to_compact.is_empty() {
+                let this_chunk_bytes: usize =
+                    chunk.iter().map(|f| f.file_size_bytes as usize).sum();
+                if this_chunk_bytes > max_total_file_size_to_group {
+                    // This chunk of files are plenty big and don't fit the ManySmallFiles charactoristics.
```
Comment (wiedld) [inline compactor/src/components/file_classifier/split_based.rs:212]: spelling nit. `charactoristics`

**Review by wiedld — APPROVED** [review]
Approved. But can we also add another test case? (even in a followup PR)

The current test case has all the larger files (the first 20 max sized files) together at the front of the `max_l0_created_at`-sorted vector. Because this sorted vector then gets chunked, and chunks whose aggregate size are too big --> those are then passed to the next round --> leaving the remaining chunks to have productive compactions.

But what happens if we have a mixture of smaller and max sized L0s across this `max_l0_created_at`-sorted vector? Such that each individual chunk may go over the limit --> even with a bunch of smaller compatible files which should be subject to `ManySmallFiles`. Is this scenario possible in the real world? If yes, then how can we also handle that?

File: compactor/tests/layouts/many_files.rs (hunk 1 above)
Comment (Joe-Blount) [reply compactor/tests/layouts/many_files.rs:4673]: without the change in this PR, it keeps using ManySmallFiles mode after it's grown the left-most L0s to be quite large.  With this PR, once 20 consecutive L0s are grown to more than the max compact size, we quit growing them more, but continue compacting/growing the rest of the small L0s.  

Another way to say it:  the issue that causes the live-lock with unproductive compactions will cause inefficiency with fewer files.  This test case has enough small files to cause the inefficiency, but not enough to cause the live-lock.

File: compactor/src/components/file_classifier/split_based.rs (hunk 2 above)
Comment (Joe-Blount) [reply compactor/src/components/file_classifier/split_based.rs:215]: I may not understand what you're getting at.

The 20 file batch of files its temporarily setting aside already total our max compact size (300MB).  On average, they're already pretty big for L0s.

This occurred in a highly backlogged partition (around 2,000 L0s).   I'm not sure why it was so backlogged, it was an older partition I noticed wasn't well compacted so I was trying to clean it up.  It appears that it takes some extenuating circumstances to hit the problem.

**Comment by Joe-Blount** [comment]
> Approved. But can we also add another test case? (even in a followup PR)
> 
> The current test case has all the larger files (the first 20 max sized files) together at the front of the `max_l0_created_at`-sorted vector. Because this sorted vector then gets chunked, and chunks whose aggregate size are too big --> those are then passed to the next round --> leaving the remaining chunks to have productive compactions.


Excellent question - but ironically most of the time & effort I spent on this issue was spent trying to get away from that test case :laughing: 

What you're asking for is approximately what I started with.  This issue was found with a real world scenario where the catalog had around 2k L0 files, many of them quite large indicating it started with much more than 2k files.  Using a test case generated from the that catalog resulted in 23k lOC.  Everything was extremely slow (IDE, compiler, running it).  

Knowing what I know now, I could write a test case with far fewer LOC, which would correct the slow compile times with the real world test.  But it would still have many more files that the included test and likely take minutes to run.  I think we'd get complaints about slowing down the tests, and requests to try to write a more focused test.  The included test case basically short cuts those huge scenarios straight to the (now avoided) live-lock problem that `ManySmallFiles` can create.

> But what happens if we have a mixture of smaller and max sized L0s across this `max_l0_created_at`-sorted vector? Such that each individual chunk may go over the limit --> even with a bunch of smaller compatible files which should be subject to `ManySmallFiles`. Is this scenario possible in the real world? If yes, then how can we also handle that?

I expect one of 3 possibilities:
1) if the scenario is too small/not enough files, then all will be handled ok
2) if the scenario approaches this live-lock problem, but doesn't get stuck, the scenario will resolve, but write amplification will be higher than normal (see the `l0s_almost_needing_vertical_split` where write amplification improved)
3) If there's enough L0 files of sufficient size, the `ManySmallFIles` algorithm will at some point create the scenario in the included test case.

**Pull request merged**